File: Jogo_da_velha.c
//DUPLA: GABRIELE MARIA FREIBERGER E BEATRICE FISCHER.

#include <stdio.h>
#include <stdlib.h>

int main (void) {
    //variáveis
    char t[3][3], jogador_atual = 'X', simbolo_jogador[3];
    int jogada_valida, jogadas_restantes, vencedor, continuar_jogo, i, j;

    simbolo_jogador[1] = 'X';
    simbolo_jogador[2] = 'O';

    do { //múltiplas partidas

        jogadas_restantes = 9;
        vencedor = 0;

        //cria o tabuleiro
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                t[i][j] = ' ';
            }
        }
        jogador_atual = 'X'; //jogador 1 comeca

        while (jogadas_restantes > 0 && vencedor == 0) {

            //mostra o tabuleiro
            printf("JOGO DA VELHA \n\n");
            printf("  0   1   2\n");
            for ( int i = 0; i < 3; i++) {
                printf("%i", i);
                for (int j = 0; j < 3; j++) {
                    printf(" %c ", t[i][j]);
                    if (j < 2) {
                        printf("|");
                    }
                }
                printf("\n");
                if (i < 2) {
                    printf(" ---|---|---\n");
                }
            }
            printf("\n----------------------\n\n");

            //validar a jogada
            jogada_valida = 0;
            while (jogada_valida == 0) {
                printf("Jogador %c, digite a linha (0 2) e coluna (0 2) colocando um espaco entre os numeros: ", jogador_atual);
				scanf("%i %i",&i,&j);
				
				//erro: numero invalido
                if ((i < 0 || i > 2) && (j < 0 || j > 2)) {
                    printf("Entrada invalida. Tente novamente.\n");
                    fflush(stdin);
                }
                //erro: caracter invalido
                if ((i < 'a' || i > 'z') && (j < 'a' || j > 'z') && (i < 'A' || i > 'Z') && (j < 'A' || j > 'Z')) {
                    printf("Entrada invalida. Tente novamente.\n");
                    fflush(stdin);
                }
                
                //erro: fora da matriz
                if (i < 0 || i > 2 || j < 0 || j > 2) {
                    printf("Posicao fora do tabuleiro. Tente novamente.\n");
                } 
                //erro: posição ocupada
                else if (t[i][j] != ' ') {
                    printf("Posicao ocupada. Tente novamente.\n");
                } 
                else {
                    jogada_valida = 1;
                    system("cls");
                }
            }

            //conta a jogada
            t[i][j] = jogador_atual;
            jogadas_restantes--;

            //verifica vitória
            for (int i = 0; i < 3; i++) {
                if (t[i][0] == t[i][1] && t[i][1] == t[i][2] && t[i][0] != ' ') {
                    if (jogador_atual == 'X') {
					    vencedor = 1;
					} else {
					    vencedor = 2;
					}
                }
                if (t[0][i] == t[1][i] && t[1][i] == t[2][i] && t[0][i] != ' ') {
                    if (jogador_atual == 'X') {
    					vencedor = 1;
					} else {
					    vencedor = 2;
					}
                }
            }
            
            if (vencedor != 0) {
            }
            // verifica a diagonal principal
            else if (t[0][0] == t[1][1] && t[1][1] == t[2][2] && t[0][0] != ' ') {
                if (jogador_atual == 'X') {
				    vencedor = 1;
				} else {
				    vencedor = 2;
				}
            }
            // verifica diagonal secundaria
            else if (t[0][2] == t[1][1] && t[1][1] == t[2][0] && t[0][2] != ' ') {
                if (jogador_atual == 'X') {
				    vencedor = 1;
				} else {
				    vencedor = 2;
				}
            }

            //troca o jogador se der velha
            if (vencedor == 0) {
             	if (jogador_atual == 'X') {
             		jogador_atual = 'O';
				 } else {
				 	jogador_atual = 'X';
				 }
            }
        }

        // tabuleiro final
        printf("\n RESULTADO \n\n");
        printf("  0   1   2\n");
        for (int i = 0; i < 3; i++) {
            printf("%i", i);
            for (int j = 0; j < 3; j++) {
                printf(" %c ", t[i][j]);
                if (j < 2) {
                    printf("|");
                }
            }
            printf("\n");
            if (i < 2) {
                printf(" ---|---|---\n");
            }
        }
        printf("\n----------------------------------------------------\n");

        //resultado do jogo
        if (vencedor == 1) {
            printf("\nParabens! O Jogador X Venceu!\n");
        } else if (vencedor == 2) {
            printf("\nParabens! O Jogador O Venceu!\n");
        } else {
            printf("\nDEU VELHA!\n");
        }
        
        // Pergunta se deseja jogar novamente
        printf("\nDeseja jogar novamente? (1 para Sim / 0 para Nao): ");
        if (scanf("%i", &continuar_jogo) != 1) {
            printf("Entrada invalida. O programa sera finalizado.\n");
            continuar_jogo = 0;
            fflush(stdin);
        }

    } while (continuar_jogo == 1);

    printf("Fim do jogo. Obrigado por jogar!\n");
    return 0;
}

File: jogo_da_velha.py
import os

# Função para limpar a tela
def limpar_tela():
    os.system('cls' if os.name == 'nt' else 'clear')

# Função para exibir o tabuleiro
def exibir_tabuleiro(tabuleiro):
    print("\nJOGO DA VELHA\n")
    print("   0   1   2")
    for i in range(3):
        print(f"{i}", end="")
        for j in range(3):
            print(f" {tabuleiro[i][j]} ", end="")
            if j < 2:
                print("|", end="")
        print()
        if i < 2:
            print(" ---|---|---")
    print("\n----------------------\n")

# Função para verificar se há um vencedor
def verificar_vitoria(tabuleiro, jogador):
    # Condições de vitória 
    vitorias = [
        # Linhas
        [tabuleiro[0][0], tabuleiro[0][1], tabuleiro[0][2]],
        [tabuleiro[1][0], tabuleiro[1][1], tabuleiro[1][2]],
        [tabuleiro[2][0], tabuleiro[2][1], tabuleiro[2][2]],
        # Colunas
        [tabuleiro[0][0], tabuleiro[1][0], tabuleiro[2][0]],
        [tabuleiro[0][1], tabuleiro[1][1], tabuleiro[2][1]],
        [tabuleiro[0][2], tabuleiro[1][2], tabuleiro[2][2]],
        # Diagonais
        [tabuleiro[0][0], tabuleiro[1][1], tabuleiro[2][2]],
        [tabuleiro[0][2], tabuleiro[1][1], tabuleiro[2][0]]
    ]
    
    # Verifica se qualquer uma das listas de vitória contém apenas o símbolo do jogador
    for linha_ou_coluna_ou_diagonal in vitorias:
        if all(simbolo == jogador for simbolo in linha_ou_coluna_ou_diagonal):
            return True
            
    return False

# Função principal do jogo
def jogar_jogo_da_velha():
    continuar_jogo = 1
    
    while continuar_jogo == 1:
        # Inicialização do tabuleiro com compreensão de lista
        tabuleiro = [[' '] * 3 for _ in range(3)] 
        jogador_atual = 'X'
        jogadas_restantes = 9
        vencedor = False

        limpar_tela()
        
        while jogadas_restantes > 0 and not vencedor:
            exibir_tabuleiro(tabuleiro)
            jogada_valida = False
            
            # Loop para obter e validar a jogada
            while not jogada_valida:
                try:
                    entrada = input(f"Jogador {jogador_atual}, digite a linha e coluna (ex: 0 2): ").split()
                    
                    if len(entrada) != 2:
                        print("Entrada inválida. Digite exatamente dois números separados por espaço.")
                        continue
                        
                    linha = int(entrada[0])
                    coluna = int(entrada[1])
                    
                    if not (0 <= linha <= 2 and 0 <= coluna <= 2):
                        print("Posição fora do tabuleiro (deve ser entre 0 e 2). Tente novamente.")
                    elif tabuleiro[linha][coluna] != ' ':
                        print("Posição ocupada. Tente novamente.")
                    else:
                        jogada_valida = True
                        limpar_tela() 
                        
                except ValueError:
                    print("Entrada inválida. Por favor, digite apenas números inteiros para linha e coluna.")

            # Executa a jogada
            tabuleiro[linha][coluna] = jogador_atual
            jogadas_restantes -= 1
            
            # Verifica vitória
            if verificar_vitoria(tabuleiro, jogador_atual):
                vencedor = True
            
            # Troca de jogador com operador ternário 
            if not vencedor:
                jogador_atual = 'O' if jogador_atual == 'X' else 'X'

        # Resultado da partida
        exibir_tabuleiro(tabuleiro)
        print("----------------------------------------------------\n")

        if vencedor:
            print(f"\nParabéns! O Jogador {jogador_atual} Venceu!\n")
        else:
            print("\nDEU VELHA!\n")
        
        # Pergunta se deseja jogar novamente
        while True:
            try:
                continuar_input = input("\nDeseja jogar novamente? (1 para Sim / 0 para Não): ")
                continuar_jogo = int(continuar_input)
                if continuar_jogo not in [0, 1]:
                    print("Opção inválida. Digite 1 ou 0.")
                    continue
                break
            except ValueError:
                print("Entrada inválida. Por favor, digite 1 ou 0.")
                
    print("\nFim do jogo. Obrigado por jogar!\n")

if __name__ == "__main__":
    jogar_jogo_da_velha()